feat: add validation of backend endpointslice

kkk777-7 opened the pull request:
**What this PR does / why we need it**:

Add validation of backend endpointslice.
If no ready endpointslices, return 500 response and report HTTPRoute/GRPCRoute status error.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #6269

<!--
For any non-trivial changes, you need to provide a brief description of the changes in the release notes.
Please add the description to the release-notes/current.yaml file and include this file in the PR.
-->
Release Notes: Yes


## Note
According to the Gateway API specification, a 503 should be returned when no valid endpoint exists, but this pr change returns 500.
https://github.com/kubernetes-sigs/gateway-api/pull/3121

Please leave a comment if this PR should be updated to return a 503 when no endpoints are available, while this may potentially change the current behavior.


Also according to [this comment](https://github.com/kubernetes-sigs/gateway-api/pull/3121#discussion_r1628840064), it seem that the `ResolvedRef` condition in xRoute might be used only for references to backend resources such as Services.

But this pr change reports an error while even though successfully retrieving the backend resource. 
Please comment if the error should not be reported.

kkk777-7 commented:
Hi @arkodg , I've fixed to set the `ResolvedRef` condition to false when there is no valid endpoint, but it seems to fail the gateway api conformance test.

> Failed to process route rule 2 backendRef 0: no ready endpoints for the related service gateway-conformance-infra/headless-manual-endpointslices.}]]
>
>    helpers.go:744: 2025-06-15T17:26:32.225529701Z: ResolvedRefs condition set to Status False with Reason EndpointSliceNotFound, expected Status True 

Checking the manifest of the failed test , I found that the endpoint of endpointslice spec was not specified.
https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/httproute-service-types.yaml#L83-L107

According to [this comment](https://github.com/kubernetes-sigs/gateway-api/pull/3121#discussion_r1628840064), it seems that ResolvedRef should not be false even if the valid endpoint of endpointslice does not exist .

---

So I'm considering the following two alternatives, which do you prefer? Do you have another approach?
- warn log is output without reflecting the status.
- define custom condition types.
```
status:
  parents:
  - conditions:
    - message: Route is accepted
      reason: Accepted
      status: "True"
      type: Accepted
    - message: Resolved all the Object references for the Route
      reason: ResolvedRefs
      status: "True"
      type: ResolvedRefs
    - message: Failed to process route rule 2 backendRef 0: no ready endpoints for the related service gateway-conformance-infra/headless-manual-endpointslices.
      reason: BackendNotFound (or EndpointSliceNotFound)
      status: "False"
      type: gateway.envoyproxy.io/BackendNotFound (or gateway.envoyproxy.io/EndpointSliceNotFound)
```

Please give your opinion when you have time.

kkk777-7 commented on internal/gatewayapi/translator.go:
```diff
@@ -103,6 +104,9 @@ type Translator struct {
 	ListenerPortShiftDisabled bool
+
+	// Logger is the logr implementation used by Envoy Gateway.
+	Logger logging.Logger
```
I’ve made changes to have the translator hold a logger, 
but please let me know if only the runner should hold a logger.

kkk777-7 commented on internal/gatewayapi/testdata/backendtrafficpolicy-buffer-limit-out-of-range-error.out.yaml:
```diff
@@ -256,6 +256,21 @@ xdsIR:
     accessLog:
       json:
       - path: /dev/stdout
+    globalResources:
```
This occurred as a result of changes made to ensure that the setup in translator_test.go works correctly.

https://github.com/envoyproxy/gateway/blob/4c5f2751391ff9aee8a14bc8dc2bfa89ca5384ec/internal/gatewayapi/translator_test.go#L375

kkk777-7 commented on internal/gatewayapi/testdata/securitypolicy-with-extauth-serviceimport.out.yaml:
```diff
@@ -213,9 +213,9 @@ securityPolicies:
-        message: Policy has been accepted.
-        reason: Accepted
-        status: "True"
+        message: 'ExtAuth: no ready endpoints for the related ServiceImport envoy-gateway/http-backend.'
```
Since ServiceImport defaults the protocol to TCP, this issue seems to be limited to the Gateway API test layer, so it shouldn’t be a problem.
https://pkg.go.dev/sigs.k8s.io/mcs-api@v0.3.0/pkg/apis/v1alpha1#ServicePort

arkodg replied:
can the test case be fixed here, or can be raise a GH issue to track it ?

arkodg commented on internal/gatewayapi/route.go:
```diff
@@ -226,11 +228,23 @@ func (t *Translator) processHTTPRouteRules(httpRoute *HTTPRouteContext, parentRe
+				// the ResolvedRefs condition should NOT be set to False.
+				// Since we cannot reflect this error in the route status, we log it so users can check the issue.
+				if err.Reason() == status.RouteReasonEndpointsNotFound {
+					t.Logger.Error(err, "failed to process route rule by no ready endpoints, so return 503 status code",
```
this is not a system error, shouldnt be logged

kkk777-7 commented:
> @kkk777-7 I think we need to flag this in the status as well because the user's use case was adding an incorrect annotation [#6269 (comment)](https://github.com/envoyproxy/gateway/issues/6269#issue-3125930061), which imo constitutes to issues resolving the reference

@arkodg thanks for review!
I agree with your opinion, and at first I implemented it so that ResolvedRef would be updated to false.
However, the Gateway API Conformance test failed, so I reverted that implementation.
Since it seems that ResolvedRef cannot be used, should we define and reflect a custom Condition instead?

arkodg commented:
also do you have any insights on why the upstream conformance test failed ?

kkk777-7 commented:
It seems that the failure occurred in the following test case:
https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/httproute-service-types.go

manifest is deployed with an empty endpoint, and then at line 90, setupEndpointSlices is called, 
which directly patches the Pod IP into the EndpointSlice.

Meanwhile, at line 75, HTTPRouteMustHaveResolvedRefsConditionsTrue is called, 
and it expects ResolvedRefs to be True even when the endpoint is empty.

With the changes in this PR, the status update caused this test case to fail.

arkodg commented:
thanks !

any suggestions for condition name ? 
https://github.com/kubernetes-sigs/gateway-api/blob/8f9ee2c6bfc0483005d4565a2e27e89235f13016/apis/v1/shared_types.go#L469

cc @envoyproxy/gateway-maintainers

kkk777-7 commented on internal/gatewayapi/testdata/httproute-with-backendref-no-endpoints.in.yaml:
This test data verifies cases where endpointslice does not exist.

kkk777-7 commented:
@arkodg 
I’ve finished the fix, could you take another look when you have a moment?

arkodg commented on internal/gatewayapi/testdata/securitypolicy-with-extauth-serviceimport.in.yaml:
```diff
@@ -70,7 +70,7 @@ serviceImports:
         - port: 80
           name: http
-          # Missing protocol here is intentional to test the default protocol handling.
+          protocol: TCP
```
curious why this was added ?
